File: backend/api/routes/quiz.py
"""Quiz API — 智慧出題 + 作答 + 歷史（roadmap 2-4e）。"""

import uuid
from typing import Any

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from api.deps import get_current_db_user, get_db
from core.errors import AppError
from core.rate_limit import rate_limit
from models.quiz import Question, QuestionType, StudentAnswer
from models.user import User
from services.quiz import (
    generate_for_student,
    generate_hint,
    list_history,
    pick_random_validated_question,
    pick_target_concept,
    submit_answer,
)
from sqlalchemy import select as sa_select  # noqa: F401  used by hint endpoint

router = APIRouter(prefix="/quiz", tags=["quiz"])


# === Schemas ===


def _mask_content_for_student(question_type: str, content: dict[str, Any]) -> dict:
    """移除答案欄位，供前端題目顯示用（避免提前洩漏）。"""
    if question_type == QuestionType.MULTIPLE_CHOICE.value:
        return {
            "stem": content.get("stem", ""),
            "options": content.get("options", []),
        }
    if question_type == QuestionType.FILL_BLANK.value:
        return {"stem": content.get("stem", "")}
    if question_type == QuestionType.CODING.value:
        return {
            "stem": content.get("stem", ""),
            "starter_code": content.get("starter_code", ""),
        }
    return {}


class GenerateRequest(BaseModel):
    type: str = Field(default=QuestionType.MULTIPLE_CHOICE.value)
    bloom_level: int = Field(default=3, ge=1, le=6)
    # 3-1e：Learn 練習 tab 指定本單元 concept；None → 走弱項補強邏輯
    concept_tag: str | None = Field(default=None, max_length=50)


class QuestionForStudentOut(BaseModel):
    """題目展示（已 mask 答案）。"""

    id: uuid.UUID
    type: str
    concept_tags: list[str]
    bloom_level: int
    difficulty: int
    content: dict

    @classmethod
    def from_question(cls, q: Question) -> "QuestionForStudentOut":
        return cls(
            id=q.id,
            type=q.type,
            concept_tags=list(q.concept_tags),
            bloom_level=q.bloom_level,
            difficulty=q.difficulty,
            content=_mask_content_for_student(q.type, q.content or {}),
        )


class SubmitRequest(BaseModel):
    question_id: uuid.UUID
    answer: dict
    time_spent_seconds: int | None = Field(default=None, ge=0)
    hint_level_used: int = Field(default=0, ge=0, le=5)


class SubmitResponse(BaseModel):
    """作答結果 + 完整解答 + LLM 解釋。"""

    answer_id: uuid.UUID  # 3-2c：供前端 fetch /quiz/answers/{id}/feedback
    is_correct: bool
    feedback: str
    correct_content: dict  # 含答案欄位（提交後才回傳）
    explanation: str


class AnswerHistoryItem(BaseModel):
    id: uuid.UUID
    question_id: uuid.UUID
    is_correct: bool
    answered_at: str
    hint_level_used: int
    time_spent_seconds: int | None
    concept_tags: list[str]


class HistoryResponse(BaseModel):
    items: list[AnswerHistoryItem]
    total: int
    page: int
    limit: int


# === Endpoints ===


@router.get("/from-bank", response_model=QuestionForStudentOut)
async def from_bank(
    concept_tag: str | None = Query(default=None, min_length=1, max_length=50),
    question_type: str | None = Query(default=None, max_length=20),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_db_user),
) -> QuestionForStudentOut:
    """Phase 6-3b：從題庫隨機抽 validated grounded 題目（不呼叫 LLM）。

    - `concept_tag` 指定 → 抽該概念題（Learn 練習 tab）
    - `concept_tag` 省略 → **U2d 弱項模式**：沿用出題 Select 邏輯挑最弱概念再抽題庫
      （Quiz 頁題庫優先，省 LLM 成本與延遲）
    - 一律排除該學生已答過的題（重複曝光防護）；全部答過 → 404 → 前端
      fallback 至 /quiz/generate 現生新題（新題 validated 後入庫，題庫自然成長）

    命中 → 直接回題目；無題 → 404 QUESTION_BANK_EMPTY，前端 fallback 至 /quiz/generate。
    """
    tag = concept_tag
    if tag is None:
        concept = await pick_target_concept(db, user.id)
        tag = concept.tag
    question = await pick_random_validated_question(
        db,
        concept_tag=tag,
        question_type=question_type,
        exclude_answered_by=user.id,
    )
    if question is None:
        raise AppError(
            404,
            "QUESTION_BANK_EMPTY",
            f"題庫尚無針對 {tag} 的可用題目（未答過且 validated）",
        )
    return QuestionForStudentOut.from_question(question)


@router.post(
    "/generate",
    response_model=QuestionForStudentOut,
    dependencies=[Depends(rate_limit("llm"))],
)
async def generate(
    body: GenerateRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_db_user),
) -> QuestionForStudentOut:
    """產生一題給當前學生。Server 自動選弱項 + 生成 + LLM 自審（retry）。

    可選指定 concept_tag（3-1e）→ 直接對該 concept 出題，跳過弱項邏輯。
    """
    question = await generate_for_student(
        db,
        user_id=user.id,
        question_type=QuestionType(body.type),
        bloom_level=body.bloom_level,
        concept_tag=body.concept_tag,
    )
    return QuestionForStudentOut.from_question(question)


@router.post("/submit", response_model=SubmitResponse)
async def submit(
    body: SubmitRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_db_user),
) -> SubmitResponse:
    """提交作答；server 判分、寫入 student_answers、更新精熟度。"""
    student_answer, question = await submit_answer(
        db,
        user_id=user.id,
        question_id=body.question_id,
        answer=body.answer,
        time_spent_seconds=body.time_spent_seconds,
        hint_level_used=body.hint_level_used,
    )
    return SubmitResponse(
        answer_id=student_answer.id,
        is_correct=student_answer.is_correct,
        feedback=student_answer.feedback,
        correct_content=question.content or {},
        explanation=question.explanation,
    )


# === Hint endpoint (3-2b) ===


class HintRequest(BaseModel):
    question_id: uuid.UUID
    hint_level: int = Field(..., ge=1, le=5)
    student_attempt: str = Field(default="", max_length=4000)


class HintResponse(BaseModel):
    level: int
    hint: str
    fallback: bool  # True = LLM 失敗用了固定句子


@router.post(
    "/hint",
    response_model=HintResponse,
    dependencies=[Depends(rate_limit("llm"))],
)
async def hint(
    body: HintRequest,
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_db_user),
) -> HintResponse:
    """為當前題目生成對應 hint_level 的提示（1-5）。

    純即時生成，不寫入 DB；學生實際使用的最高 level 由 /quiz/submit 的
    hint_level_used 欄位帶回持久化。
    """
    question = (
        await db.execute(sa_select(Question).where(Question.id == body.question_id))
    ).scalar_one_or_none()
    if question is None:
        raise AppError(404, "QUESTION_NOT_FOUND", f"找不到題目：{body.question_id}")
    if not question.validated:
        raise AppError(400, "QUESTION_NOT_VALIDATED", "此題尚未通過審查，無法取得提示")

    result = await generate_hint(
        question, hint_level=body.hint_level, student_attempt=body.student_attempt
    )
    return HintResponse(level=result.level, hint=result.hint, fallback=result.fallback)


@router.get("/history", response_model=HistoryResponse)
async def history(
    page: int = Query(default=1, ge=1),
    limit: int = Query(default=20, ge=1, le=100),
    db: AsyncSession = Depends(get_db),
    user: User = Depends(get_current_db_user),
) -> HistoryResponse:
    """取當前學生的作答歷史。"""
    rows, total = await list_history(db, user.id, page=page, limit=limit)
    items = [
        AnswerHistoryItem(
            id=r.id,
            question_id=r.question_id,
            is_correct=r.is_correct,
            answered_at=r.answered_at.isoformat(),
            hint_level_used=r.hint_level_used,
            time_spent_seconds=r.time_spent_seconds,
            concept_tags=[],  # 透過 join 補；當前 list_history 未 join，留空簡化
        )
        for r in rows
    ]
    return HistoryResponse(items=items, total=total, page=page, limit=limit)

File: web/lib/quiz.ts
/**
 * Quiz API helpers — Phase 3-1e（Learn 練習 tab 取題）。
 *
 * 對應後端 `/quiz/generate`；schema 與 `backend/api/routes/quiz.py` 一致。
 * 完整 Quiz UI（含作答提交）屬 Phase 3-2；此 lib 僅 generate 部分。
 */

import { api } from "./api";

export type QuestionType = "multiple_choice" | "fill_blank" | "coding";

export interface CodingContent {
  stem: string;
  starter_code?: string;
}

export interface MultipleChoiceContent {
  stem: string;
  options: string[];
}

export interface FillBlankContent {
  stem: string;
}

export type QuestionContent =
  | CodingContent
  | MultipleChoiceContent
  | FillBlankContent;

export interface Question {
  id: string;
  type: QuestionType;
  concept_tags: string[];
  bloom_level: number;
  difficulty: number;
  content: QuestionContent;
}

export interface GenerateQuestionPayload {
  type: QuestionType;
  bloom_level?: number;
  /** 3-1e：指定 concept 出題；省略則走後端弱項補強邏輯。 */
  concept_tag?: string;
}

export async function generateQuestion(
  payload: GenerateQuestionPayload,
): Promise<Question> {
  return api<Question>("/quiz/generate", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

/**
 * Phase 6-3b：從題庫隨機抽 validated grounded 題目（不呼叫 LLM）。
 *
 * - conceptTag 指定 → 抽該概念（Learn 練習 tab）
 * - conceptTag 省略 → U2d 弱項模式：後端沿用 Select 邏輯挑最弱概念（Quiz 頁）
 * - questionType 可選：依使用者選的題型過濾
 * - 後端一律排除已答過的題；命中 → < 1 秒回題；無題 → 拋
 *   ApiRequestError(404, "QUESTION_BANK_EMPTY")，caller fallback 至 generateQuestion。
 */
export async function getQuestionFromBank(
  conceptTag?: string,
  questionType?: QuestionType,
): Promise<Question> {
  const params = new URLSearchParams();
  if (conceptTag) params.set("concept_tag", conceptTag);
  if (questionType) params.set("question_type", questionType);
  const qs = params.toString();
  return api<Question>(`/quiz/from-bank${qs ? `?${qs}` : ""}`);
}

/**
 * K3e：以 id 直取題目（診斷嫌疑鏈微測驗入口）。
 * 僅 validated 題可取；不存在 → ApiRequestError(404, "QUESTION_NOT_FOUND")。
 */
export async function getQuestionById(questionId: string): Promise<Question> {
  return api<Question>(`/quiz/questions/${questionId}`);
}

/** 6-3c：LEARN 單元題組單題（含該學生作答狀態）。 */
export interface UnitQuestionItem {
  question: Question;
  is_answered: boolean;
  is_correct: boolean;
}

/** 6-3c：LEARN 單元題組回應。 */
export interface UnitSetResponse {
  concept_tag: string;
  items: UnitQuestionItem[];
  total: number;
  answered: number;
}

/**
 * 6-3c：取某概念的預生成題組（source='batch'）+ 作答進度。
 * LEARN 逐題作答用；不呼叫 LLM。QUIZ 弱項現生題不列入。
 */
export async function getUnitQuestionSet(
  conceptTag: string,
  questionType?: QuestionType,
): Promise<UnitSetResponse> {
  const params = new URLSearchParams({ concept_tag: conceptTag });
  if (questionType) params.set("question_type", questionType);
  return api<UnitSetResponse>(`/quiz/unit-set?${params.toString()}`);
}

/** 6-3d：弱項綜合測驗組回應。 */
export interface WeaknessSetResponse {
  questions: Question[];
  total: number;
  /** 無弱項（cold-start / 已全數掌握）→ true，前端提示先去 LEARN 練習 */
  no_weakness: boolean;
}

/**
 * 6-3d：一次生成弱項綜合測驗組（10 或 25 題）。
 * 題庫優先重用 + 缺口並行現生；單/綜合 MC 依掌握度自適應 + 1-2 綜合 coding。
 */
export async function getWeaknessSet(count: 10 | 25): Promise<WeaknessSetResponse> {
  return api<WeaknessSetResponse>(`/quiz/weakness-set?count=${count}`, {
    method: "POST",
  });
}

// === 3-2a 作答提交 ===

/** 學生作答 payload — 形狀依 question.type 決定。 */
export type SubmitAnswer =
  | { selected_index: number }      // multiple_choice
  | { code: string }                // coding
  | { answers: string[] };          // fill_blank

export interface SubmitQuestionPayload {
  question_id: string;
  answer: SubmitAnswer;
  time_spent_seconds?: number | null;
  /** 0-5；3-2b 提示系統未實作前一律 0 */
  hint_level_used?: number;
}

/** 提交後 server 回傳 — 含完整 content（含答案）+ feedback + explanation。 */
export interface SubmitResponse {
  /** 3-2c：供前端 fetch /quiz/answers/{id}/feedback */
  answer_id: string;
  is_correct: boolean;
  feedback: string;
  /** 完整題目內容（已含 answer_index / answers / 等揭露欄位）。 */
  correct_content: Record<string, unknown>;
  explanation: string;
}

export async function submitAnswer(
  payload: SubmitQuestionPayload,
): Promise<SubmitResponse> {
  return api<SubmitResponse>("/quiz/submit", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

// === 3-2b 提示系統 ===

export interface HintRequestPayload {
  question_id: string;
  hint_level: number; // 1-5
  student_attempt?: string;
}

export interface HintResponse {
  level: number;
  hint: string;
  /** true = LLM 失敗用了固定 fallback 句 */
  fallback: boolean;
}

export async function requestHint(
  payload: HintRequestPayload,
): Promise<HintResponse> {
  return api<HintResponse>("/quiz/hint", {
    method: "POST",
    body: JSON.stringify(payload),
  });
}

// === 3-2c 作答後 EDF 回饋 ===

export interface ConceptMasteryItem {
  concept_tag: string;
  concept_name_zh: string;
  /** 0.0-1.0；未練過視為 0 */
  confidence: number;
}

export interface RecommendedUnit {
  unit_id: string;
  path_id: string;
  concept_tag: string;
  concept_name_zh: string;
  video_order: number | null;
  status: "locked" | "available" | "in_progress" | "completed";
}

export interface QuizFeedbackResponse {
  concept_mastery: ConceptMasteryItem[];
  suggestion: string;
  /** true = LLM 失敗用了固定 fallback 模板 */
  suggestion_fallback: boolean;
  recommended_units: RecommendedUnit[];
}

export async function getQuizFeedback(
  answerId: string,
): Promise<QuizFeedbackResponse> {
  return api<QuizFeedbackResponse>(`/quiz/answers/${answerId}/feedback`);
}
